File: 2.68-9/scala/sol.scala
object Huffman {

  case class Tree (
    left: Option[Tree],
    right: Option[Tree],
    symbols: Array[String],
    weight: Int
  )

  def isLeaf(branch: Tree):Boolean = {
    return branch.left == None && branch.right == None
  }

  def printEncoded(message: Array[Int]): Unit = {
    message.foreach( (b) => print(b) )
    println()
  }

  def printDecoded(message: Array[String]): Unit = {
    message.foreach( (str) => print(str + " "))
    println()
  }

  def printTree(tree: Tree): Unit = {
    println("Tree : " + tree.symbols.mkString(", ") + "; " + tree.weight)
    if(!isLeaf(tree)) {
      printTree(tree.left.get)
      printTree(tree.right.get)
    }

  }

  def encodeSymbol(sym: String, tree: Tree): Array[Int] = {
    var encoded = Array.empty[Int]
    var currentBranch = tree

    while (!isLeaf(currentBranch)) {
      if(currentBranch.left.get.symbols.contains(sym)) {
        encoded = encoded :+ 0
        currentBranch = currentBranch.left.get
      }
      else {
        encoded = encoded :+ 1
        currentBranch = currentBranch.right.get
      }
    }
    println("Encoded " + sym + " as " + encoded.mkString(" "))

    return encoded
  }

  def encode(message: Array[String], tree: Tree): Array[Int] = {
    var results = Array.empty[Int]

    message.foreach( (sym) => results = results ++ encodeSymbol(sym, tree))

    return results
  }

  def decode(bits: Array[Int], tree: Tree): Array[String] = {

    var result = Array.empty[String]
    var currentBranch = tree
    var i = -1
    
    for(bit <- bits) {
      i += 1
      if(isLeaf(currentBranch)) {
        println("Decoded : " +i+" : "+ currentBranch.symbols(0))
        result = result :+ currentBranch.symbols(0)
        currentBranch = tree
      }
      if(bit == 0) {
        currentBranch = currentBranch.left.get
      }
      else {
        currentBranch = currentBranch.right.get
      }
    }

      if(isLeaf(currentBranch)) {
        println("Decoded : " +i+" : "+ currentBranch.symbols(0))
        result = result :+ currentBranch.symbols(0)
        currentBranch = tree
      }
    return result
  }

  def combineBranches(left: Tree, right: Tree): Tree = {
    return new Tree(Some(left), Some(right), left.symbols ++ right.symbols, left.weight + right.weight)
  }

  def successiveMerge(pairs: Array[Tree]): Tree = {
    var merged = pairs.clone

    while(merged.length > 1) {
      val first = merged(0)
      val second = merged(1)
      merged = merged.drop(2)
      merged = combineBranches(first, second) +: merged
      merged = merged.sortWith(_.weight < _.weight)
    }
    
    return merged.head
  }

  def generateHuffmanTree(freqList: Array[(String, Int)]): Tree = {
    var leafs = Array.empty[Tree]
    freqList.foreach( (leaf) => {leafs = new Tree(None, None, Array(leaf._1), leaf._2) +: leafs})

    leafs = leafs.sortWith(_.weight < _.weight)

    return successiveMerge(leafs)
  }

  def main(args: Array[String]): Unit = {
    var tree = generateHuffmanTree(
                        Array(("NA", 16), ("YIP", 9), ("SHA", 3), ("A", 2), 
                              ("GET", 2), ("JOB", 2), ("BOOM", 1), ("WAH", 1)))
    var encoded = encode(Array("NA", "YIP", "SHA", "JOB"), tree)
    printTree(tree)
    printEncoded(encoded)
    printDecoded(decode(encoded, tree))
  }
}

File: 2.68-9/rust/sol.rs
struct Tree {
    left: Option<Box<Tree>>,
    right: Option<Box<Tree>>,
    symbols: Vec<String>,
    weight: i32,
}

impl Tree {
    fn new_leaf(symbol: String, weight: i32) -> Tree {
        Tree {
            left: None,
            right: None,
            symbols: vec![symbol],
            weight,
        }
    }

    fn new_node(left: Tree, right: Tree) -> Tree {
        let symbols = [left.symbols.clone(), right.symbols.clone()].concat();
        let weight = left.weight + right.weight;
        Tree {
            left: Some(Box::new(left)),
            right: Some(Box::new(right)),
            symbols,
            weight,
        }
    }
}

fn is_leaf(tree: &Tree) -> bool {
    tree.left.is_none() && tree.right.is_none()
}

fn encode_symbol(symbol: &str, tree: &Tree) -> Vec<i32> {
    if is_leaf(tree) {
        return vec![];
    }

    if tree.left.as_ref().unwrap().symbols.contains(&symbol.to_string()) {
        let mut encoded = vec![0];
        encoded.extend(encode_symbol(symbol, tree.left.as_ref().unwrap()));
        encoded
    } else {
        let mut encoded = vec![1];
        encoded.extend(encode_symbol(symbol, tree.right.as_ref().unwrap()));
        encoded
    }
}

fn encode(message: Vec<&str>, tree: &Tree) -> Vec<i32> {
    let mut encoded_message = vec![];
    for symbol in message {
        encoded_message.extend(encode_symbol(symbol, tree));
    }
    encoded_message
}

fn decode(bits: Vec<i32>, tree: &Tree) -> Vec<String> {
    let mut decoded_message = vec![];
    let mut current_tree = tree;

    for bit in bits {
        if bit == 0 {
            current_tree = current_tree.left.as_ref().unwrap();
        } else {
            current_tree = current_tree.right.as_ref().unwrap();
        }

        if is_leaf(current_tree) {
            decoded_message.push(current_tree.symbols[0].clone());
            current_tree = tree;
        }
    }

    decoded_message
}

fn generate_huffman_tree(freq_list: Vec<(&str, i32)>) -> Tree {
    let mut trees: Vec<Tree> = freq_list
        .into_iter()
        .map(|(symbol, weight)| Tree::new_leaf(symbol.to_string(), weight))
        .collect();

    while trees.len() > 1 {
        trees.sort_by_key(|tree| tree.weight);
        let left = trees.remove(0);
        let right = trees.remove(0);
        let new_tree = Tree::new_node(left, right);
        trees.push(new_tree);
    }

    trees.remove(0)
}

fn main() {
    let freq_list = vec![
        ("NA", 16),
        ("YIP", 9),
        ("SHA", 3),
        ("A", 2),
        ("GET", 2),
        ("JOB", 2),
        ("BOOM", 1),
        ("WAH", 1),
    ];
    let tree = generate_huffman_tree(freq_list);

    let message = vec!["NA", "YIP", "SHA", "JOB"];
    let encoded_message = encode(message.clone(), &tree);
    let decoded_message = decode(encoded_message.clone(), &tree);

    println!("Original message: {:?}", message);
    println!("Encoded message: {:?}", encoded_message);
    println!("Decoded message: {:?}", decoded_message);
}
